### PR TITLE
Reworked dataloader, model and loss to use BCEwithlogits to get base train and validation loop working.

### PR DESCRIPTION
**Pull request by rottingbadger**

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -120,36 +128,44 @@ def train(self):
...
             loop = tqdm(self.val_loader, desc="Validating")
+
+            firstitr = True
+
             for images, labels in loop:
                 images, labels = images.to(self.device), labels.to(self.device)
 
                 outputs = self.model(images)
-                loss = self.criterion(outputs, labels)
+                loss = self.criterion(outputs.squeeze(), labels.float())
```
Comment (monezamipoor) [inline main.py:156]: This outputs.squeeze() means that there's something wrong with model architecture, this shouldn't naturally happen. We maybe need to modify model architecture and squeeze before returning logits.

File: options/default.yml [hunk 2]
```diff
@@ -31,23 +33,22 @@ training:
...
 
 # Testing and Logging Arguments
 testing:
   log_dir: "./logs"
   model_save_strategy: "best"  # Options: all, best, last, none
-  model_save_metric: ['mAP','recall']
+  model_save_metric: ['AUC']
   checkpoint_dir: "./checkpoints"
   wandb:
-    project_name: "my_pytorch_project"
```
Comment (monezamipoor) [inline options/default.yml:48]: A new TODO: Adding WandB

File: options/default.yml [hunk 3]
```diff
@@ -6,11 +6,13 @@ dataset:
...
     random_rotation: 15  # degrees
     color_jitter: 0.2  # brightness, contrast, saturation
-  oversampling_rate: 1.5  # For imbalanced datasets
-  stratified_batching: true  # For maintaining class distribution in batches
-  dataset_train_path: "./data/train"
-  dataset_train_path: "./data/validation"
-  batch_size: 32
+  #oversampling_rate: 1.5  # For imbalanced datasets
+  #stratified_batching: true  # For maintaining class distribution in batches
+  dataset_train_path: "H:/My Drive/healthdata/melanoma-256-jpg/train"
```
Comment (monezamipoor) [inline options/default.yml:11]: We need to change this to relative paths.

File: main.py [hunk 4]
```diff
@@ -120,36 +128,44 @@ def train(self):
                 loop.set_postfix(loss=loss.item())
 
             avg_loss = total_loss / len(self.train_loader)
-            val_loss, val_metrics = self.validate()
+            val_loss, val_metrics = self.validate()             #TODO Would this be better extracted outside of the train method?
```
Comment (monezamipoor) [inline main.py:131]: #TODO Would this be better extracted outside of the train method?
Yeah makes sense, definitely.

File: main.py [hunk 5]
```diff
@@ -120,36 +128,44 @@ def train(self):
...
                 images, labels = images.to(self.device), labels.to(self.device)
 
                 outputs = self.model(images)
-                loss = self.criterion(outputs, labels)
+                loss = self.criterion(outputs.squeeze(), labels.float())
                 total_loss += loss.item()
 
-                all_outputs.append(outputs.cpu())
-                all_labels.append(labels.cpu())
+                if firstitr:
```
Comment (monezamipoor) [inline main.py:159]: I didn't quite got the need for this firstitr.

File: main.py [hunk 6]
```diff
@@ -47,10 +52,10 @@ def get_scheduler(self):
         elif self.opt['training']['scheduler'] == 'reduce_on_plateau':
             return torch.optim.lr_scheduler.ReduceLROnPlateau(self.optimizer, patience=5, factor=0.1)
 
-    def freeze_backbone(self):
+    def freeze_backbone(self, freeze=False):
         for param in list(self.model.parameters())[:-1]:
-            param.requires_grad = False
-        print("Backbone layers frozen.")
+            param.requires_grad = freeze
```
Comment (monezamipoor) [inline main.py:57]: https://pytorch.org/docs/stable/generated/torch.Tensor.requires_grad_.html
It works by:
param.requires_grad = False
As mentioned in doc.

File: model.py [hunk 7]
```diff
@@ -1,55 +1,54 @@
...
-            pretrained=pretrained,
-            num_classes=0 
+            pretrained=pretrained
         )
         feature_dim = self.backbone.num_features
 
-        self.classifier = nn.Sequential(
+        # Custom head
+        self.backbone.global_pool = nn.AdaptiveAvgPool2d(1)  # Average mean pooling
+        self.backbone.fc = nn.Sequential(
```
Comment (monezamipoor) [inline model.py:23]: Handling last layer with backbone.fc, it means that while freezing using:
for param in list(self.model.parameters())[:-1]
we have to specify the exact backbone.fc layer and just indexing last layer by [:-1] is not reliable enough. 
It's doable though and it's fine. 
I'm just mentioning what needs to be changed.

File: data.py [hunk 8]
```diff
@@ -42,33 +62,50 @@ def build_transforms(self):
...
 def melanoma_dataloaders(opt):
-    
+
+    dataset = pd.read_csv(opt['dataset']['dataset_train_csv'])
+
+    files = dataset['image_name'].values + '.jpg'       # Images need .jpg to be found
+    classes = dataset['target'].values                  # Target classes (0 = benign, 1 = malignant)
+
+    # Split the dataset into 80/20 and default stratify along classes for the split. Note this does not stratify based on batches
+    train_files, val_files, train_classes, val_classes = train_test_split(files, classes, train_size=0.8,
```
Comment (monezamipoor) [inline data.py:80]: Didn't know about this. Great!

**Review by monezamipoor — APPROVED** [review]
